fix(replicationSlots): check if replication slots exist when computing required number of replicas

piotrkpc opened the pull request:
*Previously*:
- P1 (primary) and R1 and R2 (replicas)
- P1 - contains replication slots R1 and R2
-- Failover
-- P1 is down, R1 is promoted
-- R1 have 0 replication sets
-- ReplicaStatefulSet count see only **one replica R2** - expect 2
-- ReplicaStatefulSet creates replica R3
-- We have 1 primary (R1) and 2 replicas (R2 and R3)
-- But R1 only have R3 replication slot - R2 is broken.

With this fix
- P1 (primary) and R1 and R2 (replicas)
- P1 - contains replication slots R1 and R2
-- Failover
-- P1 is down, R1 is promoted
-- R1 have 0 replication sets
-- ReplicaStatefulSet count see **zero replicas** - expect 2
-- ReplicaStatefulSet creates replica R3 and R4
-- R1 (primary) have 2 replicaiton slots (R3 and R4)
-- R2 is undeployed because it don't have replication slot

piotrkpc committed:
fix(replicationSlots): check if replication slots exist when computing required number of replicas

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

piotrkpc committed:
add more standby test

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

piotrkpc committed:
export only when needed so it does not mess up with DB Con reconciler

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

sergicastro commented on controllers/spec/enforcer/resources_count_spec/statefulset/ReplicaDbCountSpecEnforcer.go:
```diff
@@ -554,6 +582,17 @@ func (r *ReplicaDbCountSpecEnforcer) hasReplicationSlotsEnabled(statefulSet stat
+				// we also need to check if the replication slot actually exist in the database
+				// if it does not exist, we should treat it as if replication slots are not enabled.
+				// This could happen after a failover when the primary changes and the new primary does not have
+				// replication slots created for existing replicas.
```
```suggestion
				// Also check if the replication slot actually exist in the database.
				// If it does not, treat it as if replication slots are not enabled, so we'll trigger a rollout of the 
				// current replica.
				// This could happen after a failover when the primary changes and the new primary does not have
				// replication slots created for the alread existing replicas.
				// Causing the rollout of the replicas will make all of them to be registered in the new primary.
```

sergicastro commented on test/primary_failure_and_recovery_test.go:
```diff
@@ -171,10 +171,68 @@ var _ = Describe("Primary instances is not available, checking recovery works",
+			test.GivenUserAddedInPrimaryDb()
+			expectedNbreUsers++
+
+			// First failover
+			test.whenPrimaryIsDeleted()
```
Before running the first failover, should we check that the current replicas have the data replicated?

piotrkpc replied:
I've added the check, but this one is equivalent to the above test

sergicastro commented on test/primary_failure_and_recovery_test.go:
```diff
@@ -171,10 +171,68 @@ var _ = Describe("Primary instances is not available, checking recovery works",
+			test.ThenReplicaDbContainsExpectedNbreUsers(expectedNbreUsers)
+
+			log.Println("waiting ")
+			//time.Sleep(10 * time.Minute)
```
We are not waiting. Remove the log or add the sleep.
If we need to wait, add a comment telling why.

sergicastro commented on test/standby_test.go:
```diff
@@ -255,6 +255,94 @@ var _ = Describe("Setting Kubegres spec 'replica'", Label("standby"), func() {
+			test.standbyDBQueryTestCases.ThenWeCanSqlQueryReplicaDb()
+
+			Eventually(func() bool {
+				replicationSlots := test.standbyDBQueryTestCases.GetReplicationSlots()
```
Question: Is this querying the replicas?

piotrkpc replied:
This is confusing but it actually queries the external database that this standby cluster following.

The other one, `activeDBQueryTestCases` allows to query a primary instance after it's `promoted`.
example: https://github.com/tetrateio/kubegres/blob/5fcbd03f4a596b28c13429fa062abfb01fa37774/test/standby_test.go#L229-L230

piotrkpc committed:
leftover code

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

piotrkpc committed:
Update controllers/spec/enforcer/resources_count_spec/statefulset/ReplicaDbCountSpecEnforcer.go

Co-authored-by: Sergi Castro <sergikstro@gmail.com>
Signed-off-by: Piotr Kopec <piotr@tetrate.io>

piotrkpc committed:
address review remarks

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

piotrkpc committed:
wrap with eventually

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

piotrkpc committed:
move test and deflake it with extra check

```golang
test.thenPodsStatesShouldBe(1, 2+2) // There should be 2 replicas left + wait for at least 2 new replicas to be created because the old replicas don't have replication slots
test.thenPodsStatesShouldBe(1, 3)   // after new replicas are created, wait for old replicas to be deleted
```

Signed-off-by: Piotr Kopec <piotr@tetrate.io>

sergicastro commented on test/primary_failure_and_recovery_test.go:
```diff
@@ -205,7 +201,8 @@ var _ = Describe("Primary instances is not available, checking recovery works",
 			test.whenPrimaryIsDeleted()
 
-			test.thenPodsStatesShouldBe(1, 3)
+			test.thenPodsStatesShouldBe(1, 2+2) // There should be 2 replicas left + wait for at least 2 new replicas to be created because the old replicas don't have replication slots
+			test.thenPodsStatesShouldBe(1, 3)   // after new replicas are created, wait for old replicas to be deleted
```
The old behavior is the correct one.
We should expect the final state here, regardless of the intermediate one.
Consider incrementing the validation timeout if required.

piotrkpc replied:
this flakyness was caused elsewhere. We started the failover without waiting for the event that reconcilation is complete and check that failover could be done safely.
I've introduced a code that selects only replica that are using replication slots if replication slots are enabled for Kubegres.
This would make the process more robust and resilient to flakyness, but IMHO it's not 100% correct because a replica without replication slots enabled is still a valid replica that should contain correct data.

piotrkpc committed:
- select replica that matches replication slots settings
- test reprobe faster - shorter retry interval
- minor logging improvement

Signed-off-by: Piotr Kopec <piotr@tetrate.io>
